### ISSUE TITLE
`carthage update Foo` fetches all dependencies, not just the tree starting at Foo

### ISSUE THREAD

**Issue by kballard** [Carthage/Carthage#1081]
I have a project with 11 direct dependencies and 16 total ones. When I run `carthage update SomeDependency`, it fetches all 16 total dependencies instead of only fetching the 2 dependencies required to handle the update for SomeDependency.

**Comment by kballard** [comment]
This is with Carthage 0.12.0.

**Comment by NachoSoto** [comment]
Yup, I can confirm this as well. It only **builds** the dependency you requested, but only after fetching all of them.

**Comment by calebd** [comment]
I could be missing something, but in 0.13 it still builds _all_ dependencies—even when a single dependency is specified.

**Comment by mdiep** [comment]
This was fixed in #1108 and will ship in the next release.

**Comment by schwa** [comment]
When is the next release?

**Comment by mdiep** [comment]
> When is the next release?

There's no specific plan, but I can do one in the next week if this is causing pain for people.

**Comment by kballard** [comment]
Please do one as soon as possible. This, and the bug where it builds everything instead of just the dependency, has already caused _hours_ of wasted time for me over the past few days.

**Comment by schwa** [comment]
Sooner the better IMO. This would dramatically improve our workflow.

**Comment by mdiep** [comment]
I've added it to my todo list and release it when I can.

In the future, feel free to open an issue requesting a new release if you're itching to get something from `master`.

**Comment by mdiep** [comment]
I've released 0.14 and [submitted a PR to homebrew](https://github.com/Homebrew/homebrew/pull/49087).

**Comment by schwa** [comment]
Fantastic. Thank you

**Comment by kballard** [comment]
@mdiep 0.14.0 did not fix this. `carthage update SomeDep` now only updates/checks out/builds the tree starting from `SomeDep`, which is good, but it still *fetches* everything.

**Comment by danroberts** [comment]
This may be a separate issue, but I have also noticed that `carthage update foo` will also update Cartfile.resolved for other dependencies.

**Comment by kballard** [comment]
@danroberts In my experience, it updates any dependencies of `foo` as well, which seems like reasonable behavior.

**Comment by cristianames92** [comment]
I want to know what happened about this issue but it is still happening. You want to update a particular dependency and its dependencies, but even if carthage updates only these ones, it clone and fetches all the dependencies, which is something that clearly can be avoided.

**Comment by guidomb** [comment]
I am also seeing this behavior, if we do `carthage update SomeDependency` all the dependencies listed on `Cartfile.resolved` are fetched.

Maybe this is expected behavior or requires a separate issue but could it be possible that if we add the `--no-use-binaries` the list of dependencies passed to `update` is ignored an all dependencies are fetched, checked-out and built?

**Comment by mdiep** [comment]
> Maybe this is expected behavior or requires a separate issue but could it be possible that if we add the --no-use-binaries the list of dependencies passed to update is ignored an all dependencies are fetched, checked-out and built?

That should not be true. Definitely open issue for that if it's what you're seeing.

**Comment by guidomb** [comment]
I'll look up on the `--no-use-binaries` issue tomorrow and let you know.

**Comment by guidomb** [comment]
@mdiep After doing some digging I realized there is NO bug with the `--no-use-binaries` flag. It was actually a problem in one of our scripts that wraps Carthage. Sorry for the false alarm.

**Comment by guidomb** [comment]
@mdiep Regarding this particular issue. I think that the problem of fetching all dependencies even when you pass the list of dependencies to updates is related to this method https://github.com/Carthage/Carthage/blob/master/Source/CarthageKit/Project.swift#L655

Could it be possible that `checkoutOrCloneDependency` should only be called for the dependencies passed to carthage command line tool (maybe also for its transitive dependencies, not sure)?

What I am saying is that instead of using the `dependencies` variables passed to the `flatMap` closure, `dependenciesToCheckout` should be used.

**Comment by mdiep** [comment]
I don't think it's that simple. I think it also needs to include the dependencies of those dependencies.

**Comment by guidomb** [comment]
Yeah I imagine that could be the case. Any hints on how to solve this issue? Is there any way to reconstruct the dependency graph from a `Cartfile.resolved` file without doing requests to GitHub?

**Comment by guidomb** [comment]
@mdiep What about using this function to get the list of dependencies and their transitive dependencies to be fetched? https://github.com/Carthage/Carthage/blob/master/Source/CarthageKit/Resolver.swift#L38

**Comment by guidomb** [comment]
Then collect the result from that method into an array of `Dependency<PinnedVersion>` and pass those to https://github.com/Carthage/Carthage/blob/master/Source/CarthageKit/Project.swift#L660. Would that work?

**Comment by mdiep** [comment]
I'm not sure. You'd need to try it out. ☺️

**Comment by racer1988** [comment]
Hello,

i have 
carthage version
0.20.1

and I can confirm that running update is always updating all the dependencies.

**Comment by patricklynch** [comment]
I am experiencing troubles because of this issue as it is described in the first post of this thread.  Any news on if this might be fixed or if there is a workaround?

**Comment by petrovi4** [comment]
Um ... maybe `carthage update`?
This command is already there, and it works as it should.
Unlike `cartage update depName`

**Comment by ZkHaider** [comment]
@petrovi4 can confirm this issue still exists for me, ideally you should be able to specify only the dependency you want to fetch/clone/build and **only** that dependency. If we need to add a new command to handle this I don't mind but is it possible to fix this @mdiep ?

**Comment by quentinfasquel** [comment]
3 years later, the issue is still around.

**Comment by blender** [comment]
@quentinfasquel  can you help fix it?

**Comment by mdiep** [comment]
PRs welcome. 🙂

Commenting here is not productive and takes time away from maintainers.
